cpp/src_prims/sparse/distance/bin_distance.cuh

```diff
@@ -0,0 +1,154 @@
+  int blocks = raft::ceildiv(n_rows * n_cols, tpb);
+  compute_jaccard_warp_kernel<<<blocks, tpb, 0, stream>>>(C, Q_norms, R_norms,
+                                                          n_rows, n_cols);
+}
+
+template <typename value_idx, typename value_t, int tpb = 1024>
+void compute_jaccard_distance(value_t *out, const value_idx *Q_coo_rows,
+                              const value_t *Q_data, value_idx Q_nnz,
+                              const value_idx *R_coo_rows,
+                              const value_t *R_data, value_idx R_nnz,
+                              value_idx m, value_idx n, cusparseHandle_t handle,
+                              std::shared_ptr<deviceAllocator> alloc,
+                              cudaStream_t stream) {
+  device_buffer<value_t> Q_norms(alloc, stream, m);
+  device_buffer<value_t> R_norms(alloc, stream, n);
+  CUDA_CHECK(
+    cudaMemsetAsync(Q_norms.data(), 0, Q_norms.size() * sizeof(value_t)));
+  CUDA_CHECK(
+    cudaMemsetAsync(R_norms.data(), 0, R_norms.size() * sizeof(value_t)));
+
+  compute_binary_row_norm_kernel<<<raft::ceildiv(Q_nnz, tpb), tpb, 0, stream>>>(
+    Q_norms.data(), Q_coo_rows, Q_data, Q_nnz);
+  compute_binary_row_norm_kernel<<<raft::ceildiv(R_nnz, tpb), tpb, 0, stream>>>(
+    R_norms.data(), R_coo_rows, R_data, R_nnz);
+
+  compute_jaccard(out, Q_norms.data(), R_norms.data(), m, n, stream);
+}
+
+/**
+ * L2 distance using the expanded form: sum(x_k)^2 + sum(y_k)^2 - 2 * sum(x_k * y_k)
+ * The expanded form is more efficient for sparse data.
+ */
+template <typename value_idx = int, typename value_t = float>
+class jaccard_expanded_distances_t : public distances_t<value_t> {
+ public:
+  explicit jaccard_expanded_distances_t(
+    const distances_config_t<value_idx, value_t> &config)
+    : config_(config),
+      workspace(config.allocator, config.stream, 0),
+      ip_dists(config) {}
+
+  void compute(value_t *out_dists) {
+    CUML_LOG_DEBUG("Computing inner products");
+    ip_dists.compute(out_dists);
+
+    value_idx *b_indices = ip_dists.trans_indices();
+    value_t *b_data = ip_dists.trans_data();
+
+    CUML_LOG_DEBUG("Computing COO row index array");
+    device_buffer<value_idx> search_coo_rows(config_.allocator, config_.stream,
+                                             config_.a_nnz);
+    csr_to_coo(config_.a_indptr, config_.a_nrows, search_coo_rows.data(),
+               config_.a_nnz, config_.stream);
+
+    CUML_LOG_DEBUG("Done.");
+
+    CUML_LOG_DEBUG("Computing Jaccard");
+    compute_jaccard_distance(out_dists, search_coo_rows.data(), config_.a_data,
+                             config_.a_nnz, b_indices, b_data, config_.b_nnz,
+                             config_.a_nrows, config_.b_nrows, config_.handle,
+                             config_.allocator, config_.stream);
+    CUML_LOG_DEBUG("Done.");
+  }
+
+  ~jaccard_expanded_distances_t() = default;
+
+ private:
+  distances_config_t<value_idx, value_t> config_;
```

Comment:
No default constructor will be called so long as `config_` is initialized by the time `jaccard_expanded_distances_t`'s constructor returns. 

However, what I don't like is that the `distance_config_t` passed into the constructor will be copied, so I will change these to be pointers.

Reply:
A reference will be easier than a pointer while achieving the same purpose. Class variables which are references are only initialized in the ctor

Reply:
> A reference will be easier than a pointer while achieving the same purpose. Class variables which are references are only initialized in the ctor

Both of these cases only initialize in the constructor- there is no default constructor called when the instance variable is not stored as a pointer, though there is a copy (e.g. a copy constructor is called, but only once).

Reply:
No copy constructor will be called if you do `distances_config_t<value_idx, value_t> &config_;`. And this will remove the complexities of using a pointer